---
title: CMDB Bootcamp 2022
layout: default
---

## CMDB Quantitative Biology Bootcamp 2022

### Logistics
- [Syllabus](syllabus/index.html)
- Five day intensive bootcamp geared towards getting everyone up to speed with basic computational skills
- <span class="text-danger">**NO** other classes, meetings, lab work, lab meetings, etc.</span>
- In person, UTL398.
- Please wear a face mask and bring headphones
- Please arrive by 8:45am ET
- Lunch and an afternoon snack will be provided
<!--- - [Join Slack](https://join.slack.com/t/jhu-cmdb/shared_invite/enQtNjk2MDQ3OTM3ODEwLTZhMzY5NjI1ZTgwYjUzZWU3Nzc4NGFhZTRkYjBjODdlN2U4YjkwZGRmYWY4YThiYmY1MGJiOGZjYmMxYTEzYjY) --->

### Schedule
All times in ET<br /><br />
All day-specific links for assignments, recordings, etc. can be found by clicking the link for the day below.


|                | Monday | Tuesday | Wednesday | Thursday |Friday |
|:---------------:|:--------------------------------------:|:---------------------------------------:|:------------------------------------:|:---------------------------------------:|:--------------------------------------:|
|     Morning      | Unix & Git | Functions & Data Structures | Algorithms | Exploratory Data Analysis | Principal Component Analysis |
|     Lunch        | Genome Arithmetic  | ID mapping | Binary Search | Data Visualization | Human Genetic Variation |
|     Afternoon    | Basic Python & Pseudocode | Writing & Running Python Scripts | Review  | Linear Models |  Review |
|     Homework     | Parsing a SAM file | Heuristic Sequence Alignment |  Assignment Catch-up   | Linear Regression | --- |
{:.table.table-striped.well}

<!--- Recall that a link looks like [Friday](day5/index.html) --->
<!--

      Basic python: running scripts, types up to list, files, reading line-by-line, split and join,
        - Exercise: summarizing SAM file, various questions not needing dictionaries
      Python II: dictionaries, functions, modules (importing) with numpy and statsmodels
        - Introduce another format and parse it, using numpy

      Python III: classes -- FASTA parser
        - Parsing blast output (too hard? we pushed this all the way to day 5 in 2015)

      Python IV: counting k-mers using FASTA parser
        - k-mer matching

-->

## CMDB Quantitative Biology Lab 2022

### Logistics
- Syllabus TBD
- Extension of Bootcamp ...
- <span class="text-danger">**NO** other classes, meetings, lab work, lab meetings, etc.</span>
- In person, Library room needs filled in.
- Please wear a face mask and bring headphones
- Time of class needs filled in.

### Schedule

TBD

### Useful Goodies

#### Unix and Git

* [Unix Reference](resources/references/unix.html)
* [Git Reference](resources/references/git.html)
* [Git Recipes](resources/references/git_recipes.html)

#### Python

* [Python and Pandas Reference](resources/references/python.html)
* [Plotting/Matplotlib Reference](resources/gallery/README.html)

#### Helpful Documentation (external links)

* [Matplotlib](http://matplotlib.org/api/pyplot_summary.html)
* [Statsmodels](http://statsmodels.sourceforge.net/stable/)
  - e.g. [ttest](http://statsmodels.sourceforge.net/stable/generated/statsmodels.stats.weightstats.ttest_ind.html)
* [NumPy](http://docs.scipy.org/doc/numpy/reference/index.html#reference)
* [Pandas](http://pandas.pydata.org/pandas-docs/stable/)
  - e.g. [read_csv](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html?highlight=read_csv)
