---
title: CMDB Quantitative Biology 2024
layout: default
---

## CMDB Quantitative Biology Bootcamp 2024

### Logistics
Prep Work: 

1. [Part 1](https://colab.research.google.com/drive/1ognkG7ph79u82DBjNoE3z9X8B4q9_ik8)
2. [Part 2](https://colab.research.google.com/drive/18K0W2S17-97AwH0BK8rog0u64QSfXbrZ )

R chapters to read before class:

1. [Workflow Basics](https://r4ds.hadley.nz/workflow-basics)
2. [Data Visualization](https://r4ds.hadley.nz/data-visualize )

- [Syllabus](resources/syllabi/bootcamp/index.html)
- Four day intensive bootcamp to introduce students to fundamentals in quantitative and computational biology 
- <span class="text-danger">**NO** other classes, meetings, lab work, lab meetings, etc.</span>
- In person, Carnegie Rose Auditorium
- Please arrive by 8:45am ET
- Lunch and an afternoon snack will be provided
<!-- - Zoom meeting ID (check Slack for passcode or full link to join): 929 0650 3074 --> 
<br /><br />
<!-- - [Answers to Student Questions](resources/references/reflection_questions_answered.html) -->

### Schedule
All times in ET<br /><br />
Resources for each day (links for assignment, any relevant recordings or slides, etc.) will be linked in the schedule below 

<!--[Tuesday](webpages/day1.md)-->
<!--[Thursday](webpages/day3.md)-->

| Time | [Tuesday](webpages/day1.md) | [Wednesday](webpages/day2.md) |  [Thursday](webpages/day3.md) | [Friday](webpages/day4.md) | 
|:---------------:|:--------------------------------------:|:---------------------------------------:|:------------------------------------:|:---------------------------------------:|
|   9:00 -  9:30  | Welcome      | Small Groups | Small Groups | Small Groups |
|   9:30 - 11:00  | Unix, git, R | Explore text files w/ Unix | Python Indexing, Arrays | Linear Regression |
|  11:00 - 12:00  | Exercise     | Exercise     | Exercise     | Exercise     | 
|  12:00 -  1:00  | Lunch        |  Lunch       | Lunch        | Lunch        |
|   1:00 -  1:30  | Small Groups | Small Groups | Small Groups | Small Groups |
|   1:30 -  3:00  | Plotting w/ ggplot2 | Parsing files w/ Python  | Python Dictionaries | Multiple testing, power analysis |
|   3:00 -  3:15  | Break        | Break        | Break        | Break        | 
|   3:15 -  5:15  | Exercise     | Exercise     | Exercise     | Exercise     | 
{:.table.table-striped.well}


<!--- Recall that a link looks like [Friday](day5/index.html) --->
<!--

      Basic python: running scripts, types up to list, files, reading line-by-line, split and join,
        - Exercise: summarizing SAM file, various questions not needing dictionaries
      Python II: dictionaries, functions, modules (importing) with numpy and statsmodels
        - Introduce another format and parse it, using numpy

      Python III: classes -- FASTA parser
        - Parsing blast output (too hard? we pushed this all the way to day 5 in 2015)

      Python IV: counting k-mers using FASTA parser
        - k-mer matching

-->

## CMDB Quantitative Biology Lab 2024

### Logistics
- [Syllabus](resources/syllabi/lab/index.html)
- This one day a week course builds upon the foundations of Quantitative Biology Bootcamp, reinforcing and expanding upon mathematical and computational methods for analysis of biological data.
- In person, Carnegie Rose Auditorium
- 10:00 - 12:30 pm, Fridays
- Assignments must be submitted within one week to your `qbb2024-answers` repository

### Schedule

| Week | Date | Instructor          | Topic                                                                             | Resubmission Deadline |
| :--: | :--: | :-----------------: | :--------------------------------------------------------------------------------:|:----: |
|  1  | 9/13 | Mike Schatz |  [Genome assembly](webpages/week1.html)| | 10/4
| 2   | 9/20 | Mike Sauria |  [Genome Content](webpages/week2.html)| | 10/13
| 3   | 9/27 | Rajiv McCoy |  [Variant discovery and genotyping](webpages/week3.html)| | 10/20
| 4   | 10/4 | Frederick Tan |  Project Work + HPC Demo | | N/A
{:.table.table-striped.well}


<!--
|  1   | 9/15  | Rajiv McCoy |  [Linear regression](webpages/week1.html) | &#124;|

|  3   | 9/29 | Dylan Taylor |  [Sequence alignment](webpages/week3.html) |&#124; |
|  4   | 10/6 | Mike Sauria |  [ChIP-seq](webpages/week4.html) |&#124; |
|  5   | 10/13 | Rajiv McCoy | [Variant calling](webpages/week5.html) |&#124; |
|  -   | 10/20 | No Class | Retreat |&#124; |
|  6   | 10/27 | Rajiv McCoy | [GWAS](webpages/week6.html) |&#124; |
|  7   | 11/3 | Mike Sauria | [Single-molecule sequencing and DNA modifications](webpages/week7.html) | 11/27 |
|  8   | 11/10 | Mike Sauria | [3D Genome](webpages/week8.html) | &#124;|
|  9   | 11/17 | Rajiv McCoy | [Bulk RNA-seq](webpages/week9.html) | 12/8 |
|  -   | 11/24 | No class | Thanksgiving Break | &#124; |
|  10   | 12/1 | Rajiv McCoy | [Data Visualization](webpages/week10.html) |	&#124; |
|  11   | 12/8 | Mike Sauria | [Single-cell RNA-seq](webpages/week11.html) |	12/15 |
-->




Resources 

### Useful Goodies

#### Unix and Git

* [Unix Reference](resources/references/unix.html)
* [Git Reference](resources/references/git.html)
* [Git Recipes](resources/references/git_recipes.html)

#### Python

* [Python and Pandas Reference](resources/references/python.html)
* [Plotting/Matplotlib Reference](resources/gallery/README.html)

#### Helpful Documentation (external links)

* [Matplotlib](http://matplotlib.org/api/pyplot_summary.html)
* [Statsmodels](http://statsmodels.sourceforge.net/stable/)
  - e.g. [ttest](http://statsmodels.sourceforge.net/stable/generated/statsmodels.stats.weightstats.ttest_ind.html)
* [NumPy](http://docs.scipy.org/doc/numpy/reference/index.html#reference)
* [Pandas](http://pandas.pydata.org/pandas-docs/stable/)
  - e.g. [read_csv](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html?highlight=read_csv)
-->

