---
title: CMDB Quantitative Biology 2022
layout: default
---

## CMDB Quantitative Biology Bootcamp 2022

### Logistics
- [Prep Work](https://kweav.github.io/prepwork_revamp/introduction.html)
- [Syllabus](resources/syllabi/bootcamp/index.html)
- Five day intensive bootcamp geared towards getting everyone up to speed with basic computational skills.
- <span class="text-danger">**NO** other classes, meetings, lab work, lab meetings, etc.</span>
- In person, UTL398.
- Please wear a face mask and bring headphones or earbuds
- Please arrive by 8:45am ET
- Lunch and an afternoon snack will be provided
- Zoom meeting ID (check Slack for passcode or full link to join): 929 0650 3074

### Schedule
All times in ET<br /><br />
All day-specific links for assignments, recordings, etc. can be found by clicking the link for the day below.


|                | [Monday](webpages/day1/index.html) | [Tuesday](webpages/day2/index.html) | [Wednesday](webpages/day3/index.html) | [Thursday](webpages/day4/index.html) | [Friday](webpages/day5/index.html) |
|:---------------:|:--------------------------------------:|:---------------------------------------:|:------------------------------------:|:---------------------------------------:|:--------------------------------------:|
|   9:00 - 9:30  | Overview                     | Homework Exercise Small Group Discussions | Homework Exercise Small Group Discussions | Homework Exercise Small Group Discussions | Homework Exercise Small Group Discussions |
|   9:30 - 11:00  | The Unix Shell | Biological Data Parsing with Python | Review | Reproducibility and GitHub | Statistical Modeling |
|   11:00 - 11:45  | Lunch    |  Lunch   | Lunch    | Lunch   | Lunch |
|   11:45 - 1:45  | Exploring Biological Data Files | Extend and Annotate a File Parser  | Data Exploration and Visualization | Research Parasites (aka How to ~~Steal~~ Reuse Other Peopleâ€™s Code) | Linear Regression |
|    1:45 - 2:00   | Break | Break | Break  | Break | Break  |
|    2:00 - 2:30   | Lunch Exercise Small Group Discussions | Lunch Exercise Small Group Discussions | Data Exploration and Visualization (cont.) | Lunch Exercise Small Group Discussions | Lunch Exercise Small Group Discussions |
|    2:30 - 4:00   | Program Documentation, Online Resources, and Debugging  | Biological Data Parsing with Python (cont.) | --- | Random Simulations  | Review |
|     Homework     | Genome Arithmetic  | Mapping across databases | Principal Component Analysis | Visualizing Simulation Results and Power | --- |
{:.table.table-striped.well}

<!--- Recall that a link looks like [Friday](day5/index.html) --->
<!--

      Basic python: running scripts, types up to list, files, reading line-by-line, split and join,
        - Exercise: summarizing SAM file, various questions not needing dictionaries
      Python II: dictionaries, functions, modules (importing) with numpy and statsmodels
        - Introduce another format and parse it, using numpy

      Python III: classes -- FASTA parser
        - Parsing blast output (too hard? we pushed this all the way to day 5 in 2015)

      Python IV: counting k-mers using FASTA parser
        - k-mer matching

-->

## CMDB Quantitative Biology Lab 2022

### Logistics
- [Syllabus](resources/syllabi/lab/index.html)
- This one day a week course builds upon the foundations of Quantitative Biology Bootcamp, reinforcing and expanding upon mathematical and computational methods for analysis of biological data.
- In person, BLC 5015
- 1 - 3:30 pm, Fridays
- Please wear a face mask and bring headphones or earbuds
- Assignments must be submitted within one week to your `qbb2022-answers` repository

### Schedule

| Week | Date | Instructor          | Topic                                                          |
| :--: | :--: | :-----------------: | :------------------------------------------------------------: |
|  1   | 9/9  | Mike Schatz (Guest) | [Genome Assembly](webpages/week1/index.html)                   |
|  2   | 9/16 | Dylan Taylor (Guest)| [Alignment](webpages/week2/index.html)                         |
|  3   | 9/23 | Rajiv McCoy         | [Variant Calling](webpages/week3)                              |
|  4   | 9/30 | Rajiv McCoy         | [GWAS](webpages/week4)                                         |
|  5   | 10/7 | Mike Sauria         | [ChIP-Seq](webpages/week5)                                     |
|  6   | 10/14| Mike Sauria         | [3D Genome](webpages/week6)                                    |
|  7   | 10/21| No class            | Retreat                                                        |
|  8   | 10/28| Mike Sauria         | Single Molecule Sequencing and DNA Modifications               |
|  9   | 11/4 | Rajiv McCoy         | Bulk RNA-Seq                                                   |
|  10  | 11/11| Kate Weaver (Guest) | Machine Learning and Applications in mRNA Expression Prediction|
|  11  | 11/18| Rajiv McCoy         | Single-Cell RNA-Seq                                            |
|  12  | 11/25| No class            | Thanksgiving Break                                             |
|  13  | 12/2 | Mike Sauria         | TBD                                                            |
|  14  | 12/9 | Rajiv McCoy         | Population Genetics                                            |
{:.table.table-striped.well}

<!--
### Useful Goodies

#### Unix and Git

* [Unix Reference](resources/references/unix.html)
* [Git Reference](resources/references/git.html)
* [Git Recipes](resources/references/git_recipes.html)

#### Python

* [Python and Pandas Reference](resources/references/python.html)
* [Plotting/Matplotlib Reference](resources/gallery/README.html)

#### Helpful Documentation (external links)

* [Matplotlib](http://matplotlib.org/api/pyplot_summary.html)
* [Statsmodels](http://statsmodels.sourceforge.net/stable/)
  - e.g. [ttest](http://statsmodels.sourceforge.net/stable/generated/statsmodels.stats.weightstats.ttest_ind.html)
* [NumPy](http://docs.scipy.org/doc/numpy/reference/index.html#reference)
* [Pandas](http://pandas.pydata.org/pandas-docs/stable/)
  - e.g. [read_csv](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html?highlight=read_csv)
-->
