<header>
<h1>
    AS.020.617
    <span style="float:right;">
        Fall 2024
    </span>
    <center>
       CMDB Quantitative Biology Lab
    </center>
</h1>
</header>

<a name="Contact"></a>
<header>
  <h2> Logistics & Contact Information</h2>
</header>

<a name="Instructors"></a>
<header><h3>Instructors</h3></header>

<table class="table table-striped" style="width:40%">
 <tr>
   <th>Instructor</th>
   <th>Email</th>
 </tr>
 <tr>
   <td>Rajiv McCoy</td>
   <td><a href="mailto:rajiv.mccoy@jhu.edu">rajiv.mccoy@jhu.edu</a></td>
 </tr>
 <tr>
   <td>Michael Sauria</td>
   <td><a href="mailto:msauria1@jhu.edu">msauria1@jhu.edu</a></td>
 </tr>
 <tr>
   <td>Frederick Tan</td>
   <td><a href="mailto:ftan2@jh.edu">ftan2@jh.edu</a></td>
 </tr>
</table>

<a name="TAs"></a>
<header><h3> Teaching Assistants </h3></header>
<table class="table table-striped" style="width:40%">
 <tr>
   <th>Teaching Assistant</th>
   <th>Email</th>
   <th>Office Hours</th>
 </tr>
 <tr>
   <td>Nicolas Moya</td>
   <td><a href="mailto:nmoya1@jh.edu">nmoya1@jh.edu</a></td>
   <td>Thursday 1-2 PM, Mudd Atrium</td>
 </tr>
 <tr>
  <td>Jonathan Fischer</td>
  <td><a href="mailto:jfisch27@jhu.edu">jfisch27@jhu.edu</a></td>
  <td>Wednesday 1-2 PM, Mudd Atrium</td>
 </tr>
 <tr>
  <td>Siqi Ma</td>
  <td><a href="mailto:sma46@jhu.edu">sma46@jhu.edu</a></td>
  <td>Monday 1-2 PM, Mudd Atrium (email me if you're there)</td>
 </tr>
</table>


<a name="ClassMeetings"></a>
<header>
  <h3>Class Meetings</h3>
</header>

Fridays, Sept 13, 2023 - Dec 6, 2024<br />
10:00 A.M. ET - 12:30 P.M. ET, Rose Auditorium, Carnegie Institution for Science<br />

<a name="CourseWebsite"></a>
<header>
  <h3>Course Website</h3>
</header>

<a href="http://bxlab.github.io/cmdb-quantbio/">http://bxlab.github.io/cmdb-quantbio/</a>

<a name="Description"></a>
<header>
  <h2>Course Description</h2>
</header>

This course builds upon the foundations of Quantitative Biology Bootcamp, reinforcing and expanding upon mathematical and computational methods for analysis of biological data. Weekly meetings of this 2.5-hour course are organized as active-learning modules focused on diverse areas of genomics. Students perform guided research with real genomic data, uploading their results and code to repositories where they receive feedback.

Topics are intergated with the curricula for concurrent core courses in molecular biology and cell biology, with datasets and analysis goals aimed at diverse topics in these fields. Examples of such topics include:

  <ul>
    <li>genome assembly, variant discovery, and genotyping</li>
    <li>analysis of functional genomic data, such as bulk and single-cell RNA-seq</li>
    <li>analysis of imaging data, including common tasks such as segmentation and classification</li>
  </ul>

Upon completing the course, students should have the background to develop reproducible bioinformatic workflows tailored to their research questions, as well as familiarity with resources for expanding upon these skills.   

<a name="Objectives"></a>
<header>
  <h2>Learning Goals</h2>
</header>
Upon completion of this course, students will be able to:
<ul>
    <li>Design workflows for common genomic data types, including developing tractable questions, organizing and preprocessing data, developing analytic pipelines, and implementing statistical models.</li>
    <li>Describe and implement common probability distributions, statistical tests, and correction for multiple hypothesis testing. Students will be able to identify confounding factors, as well as methods for their diagnosis and mitigation.</li>
    <li>Maintain a GitHub repository with code used for documentation, output, and documentation.</li>
    <li>Describe key algorithms and data structures that are commonly encountered in computational biology, along with when best to use them.</li>

  </ul>
      
<a name="StudyMats"></a>
<header>
  <h2>Study Materials</h2>
</header>
This course does not have a required text. TAs have created a [short online textbook](https://andrew-bortvin.github.io/cmdb-bootcampNotes) covering the fundamentals of Python and Git. Any lecture notes or slides will be made available on the course website. 

<a name="Format"></a>
<header>
  <h2>Course Format</h2>
</header>

<a name="Attendance"></a>
<header>
  <h3>Attendance</h3>
</header>
Due to the interactive nature of this course, there is a policy that students <b>should not participate in any other meetings, courses, or lab work during this course meeting time</b>. Please be on time for course activities, and budget extra time in anticipation of technical delays. Your attendance is expected during all synchronous sessions and your effort is also expected during asynchronous sessions. Please let us know about any emergencies, family responsibilities, illness, etc. that may prevent attendance, and we will work to accommodate reasonable requests.

<a name="Help"></a>
<header><h2>Asking for Help</h2></header>
<a name="DOIT"></a>
We both expect and encourage you to ask questions and request help throughout this course. Therefore, <b>please ask questions</b> about whatever and whenever you need to.

<a name="Googling"></a>
<header><h3>Googling</h3></header>
<b>Googling</b> is always an acceptable way to find answers or help, and we encourage you to utilize it extensively. If you adopt a solution following a Google search, make sure you understand what you incorporate, rather than just copy/pasting without comprehension of the logic or code. Please see the <a href="#Integrity">Academic Integrity & Ethics</a> syllabus section for more on this.

You may be familiar with ChatGPT and other large language models. After trying each problem/assignment/task on your own, if you’re still running into issues, feel free to use ChatGPT as you would any other online resource (Google, stack overflow, etc.). Learning how to succinctly describe exactly what you want to accomplish is a skillset in itself, so this can be good practice. If you find code that seems to work (e.g., from Google) but you’re not sure how exactly it works, you can also type it into ChatGPT and ask it to explain what’s happening. As always, please do not submit any code if you are not familiar entirely with how it works; flag it and ask a TA for assistance. Be aware that ChatGPT might confidently offer an answer that is not correct; so always check the output on your own.

<a name="Grading"></a>
<header><h2>Grading</h2></header>
Weekly exercise solutions (Jupyter notebooks, scripts, results, etc.) should be pushed to a student's `qbb2023-answers` GitHub repository within a week after the assignment is posted. The grading for this course is based on <b>reasonable completion</b>. For each weekly exercise, students will be advised which if any exercises are advanced and therefore not required for submission. <br /><br />

TAs will verify that the student’s submitted work each week shows a reasonable level of individual effort. Letter grades will be assigned in line with the level of completion.<br /><br />

General guidelines for letter grade assignments:
  <ul>
    <li>A (excellent): 90-100% completion</li>
    <li>B (passing): 80-90% completion</li>
    <li>C or lower (not passing): <80% completion</li>
  </ul>

<a name="Integrity"></a>
<header><h2>Academic Integrity & Ethics</h2></header>

Academic and scientific institutions and research depend on honesty and integrity. You should complete your own lunch and homework exercises, and your work should not plagiarize others -- including group partners, presenters, and strangers posting to online forums or blogs. You and your partners should be working together, but both persons should be writing and turning in unique, individualized code. Note your scripts and analysis may follow the same logic steps and even have tidbits of the same code, but no one person should be writing the solution the whole group uses character for character. Additionally, you should understand every line of code you write, are given and use, or find online and incorporate. If asked to, you should be able to explain exactly what your code is doing. Another aspect is properly acknowledging the source of borrowed code. Understanding can be cultivated and acknowledgement implemented by writing both inline and multiline comments (which is a terrific practice in general). Relatedly, don’t give someone code to copy and paste. Make sure any recipient can explain back to you any gifted code. See the University’s guidelines on plagiarism for details and contact a TA with any questions.

<a name="CodeofConduct"></a>
<header><h2>Code of Conduct</h2></header>

We are committed to maintaining a welcoming, inclusive, and harassment-free environment for everyone, regardless of gender, gender identity and expression, age, sexual orientation, disability, physical appearance, body size, race, ethnicity, religion (or lack thereof), political beliefs/leanings, or technology choices. We do not tolerate harassment in any form. This code of conduct applies to all course participants, including instructors and TAs, and to all modes of interaction and communication.<br /><br />

All class participants agree to:
<ul>
  <li><b>Be considerate and polite</b>. Quant Lab can be a daunting and sometimes frustrating experience - we're showing you a lot in a short amount of time! In addition, the collaborative nature of this course means it is essential that students are respectful of each other’s questions and points of confusion. Remember that you all are coming into this program with a wide range of coding backgrounds. A community where people feel comfortable is most conducive for learning.</li>
  <li><b>Work together cooperatively</b>. Be kind when giving feedback on your classmates’ code, and make an effort to help if they are stuck. You will gain much more from explaining concepts to someone else than you will from rushing through the assignments on your own.</li>
  <li><b>Refrain from demeaning, discriminatory, or harassing behavior and speech</b>. Harassment includes, but is not limited to: deliberate intimidation; stalking; unwanted photography or recording; sustained or willful disruption of talks or other events; inappropriate physical contact; use of sexual or discriminatory imagery, comments, or jokes; and unwelcome sexual attention. If you feel that someone has harassed you or otherwise treated you inappropriately, please alert one of the instructors or TAs.</li>
  <li><b>Refrain from advocating for, or encouraging, negative behavior</b>. And, if someone asks you to stop, then stop. Alert an instructor or TA if you notice a dangerous situation, someone in distress, or violations of this code of conduct, no matter how inconsequential it seems.</li>
</ul>

<a name="WeeklySchedule"></a>
<header><h2>Weekly Schedule</h2></header>

<table class="table table-striped">
<tr>
  <th>Week</th>
  <th>Date</th>
  <th>Instructor</th>
  <th>Topic</th>
</tr>
<tr>
  <td>1</td>
  <td>9/15</td>
  <td>Rajiv McCoy</td>
  <td>Linear regression</td>
</tr>
<tr>
  <td>2</td>
  <td>9/22</td>
  <td>Mike Schatz</td>
  <td>Genome assembly</td>
</tr>
<tr>
  <td>3</td>
  <td>9/29</td>
  <td>Dylan Taylor</td>
  <td>Sequence alignment</td>
</tr>
<tr>
  <td>4</td>
  <td>10/6</td>
  <td>Mike Sauria</td>
  <td>ChIP-seq</td>
</tr>
<tr>
  <td>5</td>
  <td>10/13</td>
  <td>Rajiv McCoy</td>
  <td>Variant calling</td>
</tr>
<tr>
  <td>-</td>
  <td>10/20</td>
  <td>No Class</td>
  <td>Retreat</td>
</tr>
<tr>
  <td>6</td>
  <td>10/27</td>
  <td>Rajiv McCoy</td>
  <td>GWAS</td>
</tr>
<tr>
  <td>7</td>
  <td>11/3</td>
  <td>Mike Sauria</td>
  <td>Single-molecule sequencing and DNA modifications</td>
</tr>
<tr>
  <td>8</td>
  <td>11/10</td>
  <td>Mike Sauria</td>
  <td>3D Genome</td>
</tr>
<tr>
  <td>9</td>
  <td>11/17</td>
  <td>Rajiv McCoy</td>
  <td>Bulk RNA-seq</td>
</tr>
<tr>
  <td>-</td>
  <td>11/24</td>
  <td>No class</td>
  <td>Thanksgiving Break</td>
</tr>
<tr>
  <td>10</td>
  <td>12/1</td>
  <td>Rajiv McCoy</td>
  <td>Metagenomics</td>
</tr>
<tr>
  <td>11</td>
  <td>12/8</td>
  <td>Mike Sauria</td>
  <td>Single-cell RNA-seq</td>
</tr>
</table>
