# Lab Week 10 - Machine Learning and Applications in mRNA Expression Prediction

## Lecture -- Kate Weaver (Guest)

[Lecture Slides](https://github.com/bxlab/cmdb-quantbio/raw/main/assignments/lab/machine_learning/slides_asynchronous_or_livecoding_resources/machine_learning_slides.pdf)

[Recording of lecture](https://livejohnshopkins.sharepoint.com/:v:/s/qbb2021/EQaI0YBtkTVCk2vHORF9GDsBdweEP31C9Cl6pTpaQkxrSg?e=InMVd7)

## Live-coding

[Xpresso data website](https://xpresso.gs.washington.edu/data/)

Live-coding master script: `~/cmdb-quantbio/assignments/lab/machine_learning/slides_asynchronous_or_livecoding_resources/parse_xpresso_predictions.py`

## Homework Assignment

Please upload your live-coding script/plot for credit for this week's assignment.

[Homework assignment -- optional](https://bxlab.github.io/cmdb-quantbio/assignments/lab/machine_learning/assignment/)

## Applied Paper Example

Please read the paper ["Predicting mRNA Abundance Directly from Genomic Sequence Using Deep Convolutional Neural Networks"](https://pubmed.ncbi.nlm.nih.gov/32433972/) before class. Focus specifically on the following sections and figures:

  * Introduction
  * Figure 1
  * Results section "An Optimized Deep-Learning Model to Predict mRNA Expression Levels"
  * Figure 2
  * Results section "Performance of Predicitve Models in Human and Mouse"
  * Figure 3
  * Results section "Cell-Type-Specific Models Implicate a Diversity of Gene Regulatory Mechanisms"
  * Figure 6
  * Results section "Xpresso Highlights the Importance of the Core Promoter Sequence in Predicting Expression Levels"
  * Results section "CpG Dinucleotides Are the Dominant Signal Explaining Expression Levels"
